src/messages.c

```diff
@@ -0,0 +1,47 @@
+// NOTE: This file has been generated automatically by “update-message-registry.py”.
+//       Do not edit manually!
+
+#include <stdlib.h>
+#include "messages.h"
+
+static const struct xkb_message_entry xkb_messages[] = {
+    {XKB_WARNING_UNRECOGNIZED_KEYSYM},
+    {XKB_WARNING_CANNOT_INFER_KEY_TYPE},
+    {XKB_WARNING_UNDEFINED_KEY_TYPE},
+    {XKB_WARNING_NON_BASE_GROUP_NAME},
+    {XKB_WARNING_CONFLICTING_KEY_SYMBOL},
+    {XKB_WARNING_EXTRA_SYMBOLS_IGNORED},
+    {XKB_WARNING_MULTIPLE_GROUPS_AT_ONCE},
+    {XKB_ERROR_INVALID_SYNTAX},
+    {XKB_WARNING_UNDEFINED_KEYCODE},
+    {XKB_WARNING_CONFLICTING_MODMAP},
+    {XKB_WARNING_CONFLICTING_KEY_ACTION},
+    {XKB_WARNING_CONFLICTING_KEY_TYPE},
+    {XKB_WARNING_CONFLICTING_KEY_FIELDS},
+    {XKB_WARNING_UNRESOLVED_KEYMAP_SYMBOL},
+    {0}
+};
+
+const struct xkb_message_entry*
+xkb_message_get(xkb_message_code_t code)
+{
+    size_t first = 0;
+    size_t last = sizeof(xkb_messages) / sizeof(*xkb_messages) - 2;
+
+    if (code < xkb_messages[0].code || code > xkb_messages[last].code)
+        return NULL;
+
+    /* binary search in xkb_messages */
+    while (last >= first) {
+        size_t mid = (first + last) / 2;
+        if (xkb_messages[mid].code < code)
+            first = mid + 1;
+        else if (xkb_messages[mid].code > code)
+            last = mid - 1;
+        else /* found it */
+            return &xkb_messages[mid];
+    }
```

Comment:
my absolute guess is that with the handful of messages we have, simply iterating through the list until found or `code > current` would only be fractions of a milliseconds but make the code simpler and more robust to the few edge cases. 

As ridiculous as this sounds I'd like to see a test *somewhere* to make sure this function works with the various edge cases. But it looks correct and it's not in a critical path, so this can be done as follow-up.

Reply:
Done